src/cmap/worker.rs

```diff
@@ -0,0 +1,494 @@
+        options: Option<ConnectionPoolOptions>,
+    ) -> (PoolManager, ConnectionRequester) {
+        let establisher = ConnectionEstablisher::new(http_client, options.as_ref());
+        let event_handler = options.as_ref().and_then(|opts| opts.event_handler.clone());
+
+        // The CMAP spec indicates that a max idle time of zero means that connections should not be
+        // closed due to idleness.
+        let mut max_idle_time = options.as_ref().and_then(|opts| opts.max_idle_time);
+        if max_idle_time == Some(Duration::from_millis(0)) {
+            max_idle_time = None;
+        }
+
+        let max_pool_size = options
+            .as_ref()
+            .and_then(|opts| opts.max_pool_size)
+            .unwrap_or(DEFAULT_MAX_POOL_SIZE);
+
+        let min_pool_size = options.as_ref().and_then(|opts| opts.min_pool_size);
+
+        let connection_options: Option<ConnectionOptions> = options
+            .as_ref()
+            .map(|pool_options| ConnectionOptions::from(pool_options.clone()));
+
+        let (handle_listener, handle) = HandleListener::new();
+        let (connection_requester, request_receiver) =
+            ConnectionRequester::new(address.clone(), handle);
+        let (manager, management_receiver) = PoolManager::new();
+
+        let worker = ConnectionPoolWorker {
+            address: address.clone(),
+            event_handler: event_handler.clone(),
+            max_idle_time,
+            min_pool_size,
+            establisher,
+            next_connection_id: 1,
+            total_connection_count: 0,
+            generation: 0,
+            connection_options,
+            available_connections: VecDeque::new(),
+            max_pool_size,
+            request_receiver,
+            management_receiver,
+            manager: manager.clone(),
+            unfinished_check_out: None,
+            handle_listener,
+        };
+
+        RUNTIME.execute(async move {
+            worker.execute().await;
+        });
+
+        (manager, connection_requester)
+    }
+
+    /// Run the worker thread, listening on the various receivers until all handles have been
+    /// dropped. Once all handles are dropped, the pool will close any available connections and
+    /// emit a pool closed event.
+    async fn execute(mut self) {
+        let mut maintenance_interval = RUNTIME.interval(Duration::from_millis(500));
+
+        loop {
+            // If there's an outstanding request and we can serve it, do so. Otherwise poll
+            // for a new request
+            let task = if self.unfinished_check_out.is_some()
+                && self.can_service_connection_request()
+            {
+                PoolTask::CheckOut(self.unfinished_check_out.take().unwrap())
+            } else {
+                tokio::select! {
+                    Some(result_sender) = self.request_receiver.recv(),
+                                        if self.unfinished_check_out.is_none() => PoolTask::CheckOut(result_sender),
```

Comment:
I could be wrong, but this looks like if the pool is empty, checkouts will be delayed until the next tick of the maintenance interval populates it. If so, this should be fixed so that a new connection created immediately. Am I understanding this correctly?

Reply:
I don't think that is the case. If a check out request is received and the pool is empty, this branch will be taken and `check_out` will be called, which conditionally creates a connection as necessary.

Reply:
Okay, when reading all this over the second time, I think I understand what's going on better.